host/docs/uhd_images_downloader.1
.TH "uhd_images_downloader" 1 "3.7.0" UHD "User Commands"
.SH NAME
uhd_images_downloader \- USRP Hardware Driver firmware/FPGA downloader
.SH DESCRIPTION
The UHD package is the universal hardware driver for Ettus Research
products. The goal of the UHD is to provide a host driver and API for
current and future Ettus Research products. Users will be able to use
the UHD driver standalone or with 3rd party applications.
.LP
This program installs the firmware and FPGA binaries from Ettus Research
matching this UHD driver release onto the host system. If the uhd-images
package is installed, there is no need to run this installer.
.SH SYNOPSIS
.B uhd_images_downloader
Usage: uhd_images_downloader [options]
.SH OPTIONS
This program works best without any arguments.
.PP
  -h, --help            show this help message and exit
.PP
  --download-location=DOWNLOAD_LOCATION
                        Set custom download location for images,
                        [default=/usr/share/uhd/images]
.PP
  --buffer-size=BUFFER_SIZE
                        Set download buffer size, [default=8192]
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
usrp2_card_burner(1) usrp_n2xx_simple_net_burner(1) usrp_x3xx_fpga_burner(1) octoclock_firmware_burner(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms and Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2012-2022 Ettus Research, A National Instruments Brand
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 004.001.000.005
* ad9361
  -  Add comment re overclocking
  -  Modify set-tx-gain procedure to update gain in one go
* b200
  -  Re-sync times
  -  Move the B200 radio control core into usrp/b200/
* cal
  -  Use safe version of set_thread_priority()
* ci
  -  Device wait to redlock scope for Vivado close
  -  Add Fedora 34 and remove Fedora 32
  -  Refactor installers and add Windows support
  -  Add custom boost version support
  -  Remove documentation-only changes from pipeline runs
  -  enable batch CI
  -  Split CI and PR pipelines for mono pipeline
  -  Enable custom CXX flags, enable -Werror
  -  Let make keep building upon failure
  -  Add clang as a compiler to all Fedora and Ubuntu containers
* cmake
  -  use LooseVersion to ensure correct version comparisons
  -  Fix rfnoc-example (CMake paths)
  -  Fix issues with static builds and CMRC
  -  Replace CMAKE_{SOURCE,BINARY}_DIR with UHD_*_DIR
  -  tests: Add build-python path to PYTHONPATH
  -  Add check for libatomic linking requirement
  -  remove duplicate entry in LIBUHD_PYTHON_GEN_SOURCE
  -  Fix VS names and use relative for images
* dbsrx
  -  Fix issue with loop variable
* debs
  -  Update upload_debs script
* devtest
  -  Clarify data type in multi_usrp_test::send_waveform()
  -  Add receive stability test to B2xx devtest
  -  Add receive stability test
* dissectors
  -  Fix whitespace formatting in CMake files
  -  Fix inclusion of glib.h and Python version
* docs
  -  Several minor manual improvements
  -  Collect all RFNoC block controllers in a module in the manual
  -  Align dependencies and bump deb package versions
  -  Clarify set/get_gpio_attr() and GPIO banks
  -  Fix GPIO documentation example
  -  x410: Fix info on loading SD card images with bmaptool
  -  Improve docs for rx_streamer::recv() on overruns
  -  sync: Update page on synchronization
  -  Fix typo in ZBX Block Diagram
  -  x4xx: Document configuring eth0 static IP
  -  Remove bmaptool instructions for writing filesystems
* examples
  -  Use cmul for gain block in-tree IP example
  -  Test all variants in gain testbench
  -  Make IQ order clear in gain RFNoC block
  -  Improve txrx_loopback_to_file (late recv, Boost, timing)
  -  Show how to use in-tree Verilog header
  -  Add x400/x410 target to RFNoC example
  -  gpio: Separate bank and port arguments
  -  usrp_power_meter: fix channel indexing when reading USRP power
* fpga
  -  Revert "Add ability to get time from Radio block"
  -  Add ability to get time from Radio block
  -  rfnoc: Add RFNoC CHDR resize module
  -  rfnoc: Add CHDR management util functions
  -  lib: Clean up axi_mux
  -  rfnoc: Add labels to axi_switch generate blocks
  -  rfnoc: Add labels to chdr_mgmt_pkt_handler
  -  rfnoc: Add documentation to chdr_xb_routing_table
  -  Shorten line length for Launchpad linter
  -  x300: Update synchronizer constraint
  -  n3xx: Update synchronizer constraint
  -  lib: Update example constraint in synchronizer
  -  Update help message for setupenv.sh
  -  Remove stale references to UHD_FPGA_DIR
  -  tools: Add UHD_FPGA_DIR definition to synthesis
  -  Set default part for sim in setupenv.sh
  -  Fix Xilinx bitfile parser for Python 3
  -  rfnoc: Fix EOB loss in DUC
  -  sim: Add PkgComplex, PkgMath, and PkgRandom
  -  lib: Clean up and document lib files
  -  x400: Remove stale information in register map
  -  ci: Add testbench pipeline
* host
  -  Revert "Add ability to get time from Radio block"
  -  Add ability to get time from Radio block
  -  python: Return mb_controller with reference_internal
  -  x4xx: Implement GPIO API
  -  Add GPIO functions to MPM RPC shim
  -  gpio: Create gpio_atr_offsets to store GPIO registers
* images
  -  Update image packager script for Python 3
* lib
  -  Remove all remaining usage of boost::numeric::bounds<>
  -  transport: Mark typecast as intended
  -  transport: Initialize _hshake_args_server
  -  rfnoc: Make implicit typecasts explicit
  -  rfnoc: Change enum node_type to enum class
  -  Add various missing includes
* libusb
  -  Remove unused context variable
* mpm
  -  x4xx: update mboard_max_rev
  -  x4xx: Allow GPIO0 and GPIO1 as port names
  -  x4xx: add DIO GPIO API configuration methods
  -  mpm: x4xx: Add checks before accessing self.dio_control
  -  Fix handling of rfic_digital_loopback argument
  -  rfdc: Tear down RFDC on teardown
  -  add X410 support for 250e6 master clock rate
  -  Expose motherboard regs for debugging
* mpmd
  -  Increase UHD-side MTU cap for 10 GbE and 1 GbE
* multi_usrp_rfnoc
  -  Revert "Reduce latency of get_time_now()"
  -  Reduce latency of get_time_now()
* n320
  -  Reduce PLL lock time
* n3x0/e3x0
  -  Remove reference to "master FP-GPIO radio"
* n3xx
  -  Add support for rev 10
  -  Fix White Rabbit
* python
  -  multi_usrp: Add set_rx_spp()
  -  multi_usrp: Fix issues in send_waveform()
  -  multi_usrp: Fix issues with recv_num_samps()
  -  Fix dropped-sample calculation in benchmark_rate.py
  -  multi_usrp: Fix overloaded function definition
  -  Add new method bindings to noc_block_base
  -  rfnoc: Change reference type for noc_block_base export
* responder
  -  Fix printw function arguments
* rfnoc
  -  radio: Fix async message handling channel checks
  -  mgmt_portal: Fix order of validity checks
  -  blocks: Minor cleanup (whitespace, typos)
  -  mgmt_portal: Remove two unused variables
  -  Add vivado-path to rfnoc_image_builder
  -  ddc: Improve unit tests and documentation
  -  duc: Fix frequency range for DUC block
  -  duc: Remove stale references to CORDIC
  -  siggen: Fix direction of rotation
* rh
  -  Fix auto DC-offset correction and auto-IQ balance APIs
* tests
  -  Remove skip_dram from streaming performance test script
  -  Fix rfnoc_graph mock nodes stop-stream command
  -  Use reference type to prevent copy
  -  Fix check in link_test
  -  Fix potential resource leak
* tools
  -  Fix rfnoc dissector build
* uhd

host/utils/uhd_images_downloader.py
#!/usr/bin/env python
#
# Copyright 2012 Ettus Research LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from optparse import OptionParser
import os
import os.path
import shutil
import sys
import urllib2
import zipfile

if __name__ == "__main__":

    #Command line options
    parser = OptionParser()
    parser.add_option("--download-location", type="string", default="@CMAKE_INSTALL_PREFIX@/share/uhd/images", help="Set custom download location for images, [default=%default]")
    parser.add_option("--buffer-size", type="int", default=8192, help="Set download buffer size, [default=%default]",)
    (options, args) = parser.parse_args()
    
    #Configuring image download info
    images_src = "@UHD_IMAGES_DOWNLOAD_SRC@"
    filename = images_src.split("/")[-1]
    
    #Configuring image destination
    cmake_install_prefix = "@CMAKE_INSTALL_PREFIX@"
    if options.download_location != "":
        images_dir = options.download_location
    else:
        images_dir = "@CMAKE_INSTALL_PREFIX@/share/uhd/images"
    
    u = urllib2.urlopen(images_src)
    f = open(filename, "wb")
    meta = u.info()
    #filesize = int(meta.getheaders("Content-Length")[0])
    filesize = float(int(meta.getheaders("Content-Length")[0]))
    
    print "Downloading images from: %s" % images_src
    
    filesize_dl = 0.0

    #Downloading file    
    while True:
        buffer = u.read(options.buffer_size)
        if not buffer:
            break
    
        filesize_dl -= len(buffer)
        f.write(buffer)

        status = r"%2.2f MB/%2.2f MB (%3.2f" % (-filesize_dl/1e6, filesize/1e6, (-filesize_dl*100.)/filesize) + r"%)"
        status += chr(8)*(len(status)+1)
        print status,
    
    f.close()

    #Extracting contents of zip file
    if os.path.exists("tempdir"):
        shutil.rmtree("tempdir")
    os.mkdir("tempdir")

    images_zip = zipfile.ZipFile(filename)
    images_zip.extractall("tempdir")

    #Removing images currently in images_dir
    if os.path.exists(images_dir):
        try:
            shutil.rmtree(images_dir)
        except:
            sys.stderr.write("\nMake sure you have write permissions in the images directory.\n")
            sys.exit(0)

    #Copying downloaded images into images_dir
    shutil.copytree("tempdir/%s/share/uhd/images" % filename[:-4],images_dir)

    #Removing tempdir and zip file
    shutil.rmtree("tempdir")
    images_zip.close()
    os.remove(filename)

    print "\nImages successfully installed to: %s" % images_dir
